From 15dc0790111d386dd8fe455eea0857afb3f94529 Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Wed, 23 Aug 2017 15:02:20 +0800
Subject: [PATCH 2444/5242] MLK-16255-1 video: fbdev: dcss: add 'db' support
 for ctxld

commit  beddccb1db4bf9ebee70ea185c3578fc6166e1dd from
https://source.codeaurora.org/external/imx/linux-imx.git

Add double buffer Context Loader function support in
DCSS, since some of the submodules have shadow reigsters
which are more suitable to use double buffer context
loader to load the reigster values for them, and double
buffer loadding is earlier than single buffer loading.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/imx_dcss.c |    8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/video/fbdev/mxc/imx_dcss.c b/drivers/video/fbdev/mxc/imx_dcss.c
index 89898c2..11a427d 100644
--- a/drivers/video/fbdev/mxc/imx_dcss.c
+++ b/drivers/video/fbdev/mxc/imx_dcss.c
@@ -673,7 +673,6 @@ static void fill_sb(struct cbuffer *cb,
 	cb->sb_data_len++;
 }
 
-#if 0
 static void fill_db(struct cbuffer *cb,
 		    uint32_t offset,
 		    uint32_t value)
@@ -692,7 +691,6 @@ static void fill_db(struct cbuffer *cb,
 	fill_unit(unit, offset, value);
 	cb->db_data_len++;
 }
-#endif
 
 static void ctxld_fifo_info_print(struct ctxld_fifo *cfifo)
 {
@@ -2131,9 +2129,11 @@ static void dcss_ctxld_config(struct work_struct *work)
 		       info->base + chans->ctxld_addr + CTXLD_SB_COUNT);
 	}
 
+	/* configure db buffer */
 	if (cc->db_data_len) {
-		writel(cfifo->dma_handle + cfifo->sgl[0].offset +
-		       cc->sb_data_len * kfifo_esize(&cfifo->fifo),
+		writel(cfifo->dma_handle +
+		       (cc->fifo_in + cc->sb_data_len) *
+		       kfifo_esize(&cfifo->fifo),
 		       info->base + chans->ctxld_addr + CTXLD_DB_BASE_ADDR);
 		writel(cc->db_data_len,
 		       info->base + chans->ctxld_addr + CTXLD_DB_COUNT);
-- 
1.7.9.5

